### PR TITLE
[Streams] Use new Twitch API and Bearer tokens

### LINKED ISSUE #3479: Passing something with `/` to `[p]streamalert twitch` gives a very unfriendly error
# Command bugs

#### Command name

`[p]streamalert twitch bad/arg`

#### What cog is this command from?

Streams

#### What were you expecting to happen?

Give a friendly error on such fail.

#### What actually happened?

I got error "The Twitch token is either invalid or has not been set. See `=streamset twitchtoken`."

#### How can we reproduce this issue?

1. Run `[p]streamalert twitch bad/arg`
2. See error.

// I thought someone made an issue for that already but it doesn't look like it.

### LINKED ISSUE #3489: Feedback to try to help fix Streamalerts
# Command bugs

#### Command name

Streamalerts

#### What cog is this command from?

Base

#### What were you expecting to happen?

Streamalerts to work

#### What actually happened?

Streamalerts stop working

#### How can we reproduce this issue?

Posting some feedback on the twitch streamalerts (sorry that this is kind of my issue, I am involved with a quite a few servers where they are connected through twitch streaming).

It seems for servers and/or streams that I did not set an alert for before upgrading, it works fine for a few times.

However, when trying to toggle the old streamalerts (stopping and starting again) or anything like that, they just never pop.

Not sure if this helps or if this is the known mechanism, but trying to contribute what I can to try to resolve this, otherwise, I will need to try to find some other bot that does the job in the meantime =(.

I haven't been able to fully test to know how many times it works before it stops.

In case you want to know error reports, there really haven't been many:

Hope this helps. Please fix stream alerts =(.

[2020-01-25 15:31:30] [INFO] red.audio.manager: Internal Lavalink server started. PID: 6589
[2020-01-25 22:39:51] [CRITICAL] red.main: Caught unhandled exception in task: Task was destroyed but it is pending!
[2020-01-25 22:39:51] [CRITICAL] red.main: Caught unhandled exception in task: Task was destroyed but it is pending!
Exception ignored in: <generator object WebSocketCommonProtocol.close_connection at 0x1127cb270>
Traceback (most recent call last):
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/client.py", line 101, in __init__
    super().__init__(coro, loop=loop)
RuntimeError: generator ignored GeneratorExit
[2020-01-27 10:43:20] [INFO] red.cleanup: KTCAOP(87304680073351168) deleted 2 messages in channel general.
[2020-01-27 10:52:37] [INFO] red.cleanup: KTCAOP(87304680073351168) deleted 10 messages in channel teh_input.
[2020-01-28 11:58:37] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-28 14:10:27] [INFO] red.cleanup: KTCAOP(87304680073351168) deleted 3 messages in channel lolers.
[2020-01-28 14:39:20] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-28 16:33:36] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-29 04:34:12] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-29 05:06:11] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-29 06:20:34] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-29 15:13:13] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-29 17:26:48] [INFO] red.cleanup: KTCAOP(87304680073351168) deleted 5 messages in channel chefpk-feed.
[2020-01-29 17:27:07] [INFO] red.cleanup: KTCAOP(87304680073351168) deleted 2 messages in channel chefpk-feed.
[2020-01-30 08:29:42] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-30 09:38:27] [INFO] red.cleanup: Zagric(132920409447202818) deleted 20 messages in channel teh_gulag.
[2020-01-30 09:38:41] [INFO] red.cleanup: Zagric(132920409447202818) deleted 13 messages in channel teh_gulag.
[2020-01-30 14:46:08] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-31 00:01:16] [ERROR] discord.client: Attempting a reconnect in 0.16s
Traceback (most recent call last):
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/protocol.py", line 528, in transfer_data
    msg = yield from self.read_message()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/protocol.py", line 580, in read_message
    frame = yield from self.read_data_frame(max_size=self.max_size)
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/protocol.py", line 645, in read_data_frame
    frame = yield from self.read_frame(max_size)
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/protocol.py", line 706, in read_frame
    frame = yield from Frame.read(
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/framing.py", line 100, in read
    data = yield from reader(2)
  File "/Users/ktcaop/.pyenv/versions/3.8.1/lib/python3.8/asyncio/streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "/Users/ktcaop/.pyenv/versions/3.8.1/lib/python3.8/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
ConnectionResetError: [Errno 54] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/gateway.py", line 468, in poll_event
    msg = await self.recv()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/protocol.py", line 350, in recv
    yield from self.ensure_open()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/websockets/protocol.py", line 500, in ensure_open
    raise ConnectionClosed(
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/shard.py", line 78, in poll
    await self.ws.poll_event()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/gateway.py", line 473, in poll_event
    raise ResumeWebSocket(self.shard_id) from exc
discord.gateway.ResumeWebSocket: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/client.py", line 457, in connect
    await self._connect()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/shard.py", line 274, in _connect
    f.result()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/shard.py", line 84, in poll
    self.ws = await asyncio.wait_for(coro, timeout=180.0, loop=self.loop)
  File "/Users/ktcaop/.pyenv/versions/3.8.1/lib/python3.8/asyncio/tasks.py", line 483, in wait_for
    return fut.result()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/gateway.py", line 224, in from_client
    gateway = await client.http.get_gateway()
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/http.py", line 796, in get_gateway
    data = await self.request(Route('GET', '/gateway'))
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/discord/http.py", line 158, in request
    async with self.__session.request(method, url, **kwargs) as r:
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/aiohttp/client.py", line 1012, in __aenter__
    self._resp = await self._coro
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/aiohttp/client.py", line 504, in _request
    await resp.start(conn)
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/aiohttp/client_reqrep.py", line 847, in start
    message, payload = await self._protocol.read()  # type: ignore  # noqa
  File "/Users/ktcaop/redenv/lib/python3.8/site-packages/aiohttp/streams.py", line 591, in read
    await self._waiter
aiohttp.client_exceptions.ClientOSError: [Errno 54] Connection reset by peer
[2020-01-31 00:02:02] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-31 00:02:40] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-31 02:53:32] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.
[2020-01-31 13:38:42] [WARNING] discord.gateway: Can't keep up, websocket is 29.7s behind.
[2020-01-31 21:23:44] [WARNING] discord.gateway: Shard ID 0 has stopped responding to the gateway. Closing and restarting.

### PR DESCRIPTION
**Pull request by PredaaA**
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
- Stop using Twitch API v5 and use the "new API" instead.
- Use Bearer tokens for Twitch requests. Which gives 800 req/min instead of 30 when don't using them. They are stored in memory and re-generated when their expiry date come (~60 days).
- Escape markdown and mass mention for "streamer_name is live!" messages.
- Use humanize_number for every numbers.

**Important note:** That PR is based of #3237 and need that one merged before.
Fixes #3479.
Fixes #3489.

### CODE REVIEW

File: redbot/cogs/streams/streams.py [hunk 1]
```diff
@@ -92,16 +104,62 @@ def check_name_or_id(self, data: str):
...
+            except KeyError:
+                prefix = (await self.bot._config.prefix())[0]
+                message = _(
+                    "You need a client secret key to use correctly Twitch API on this cog.\n"
+                    "Follow these steps:\n"
+                    "1. Go to this page: https://dev.twitch.tv/console/apps.\n"
+                    '2. Click "Manage" on your application.\n'
+                    '3. Click on "New secret".\n'
+                    "5. Copy your client ID and your client secret into:\n"
+                    "6. Run the command `{prefix}set api twitch client_id <your_client_id_here> "
```
Comment (TrustyJAID) [inline redbot/cogs/streams/streams.py]: This should be changed to 
```suggestion
                    "`{prefix}set api twitch client_id <your_client_id_here> "
```

**Review by TrustyJAID — CHANGES_REQUESTED** [review]

File: redbot/cogs/streams/streams.py (hunk 1 above)
Comment (PredaaA) [reply redbot/cogs/streams/streams.py]: Done.

**Review by TrustyJAID — DISMISSED** [review]
This has been working well for martine for over a week now and looks like a solid addition considering Twitch will be requiring OAuth for all tokens on [April 30, 2020](https://discuss.dev.twitch.tv/t/requiring-oauth-for-helix-twitch-api-endpoints/23916). While we weren't using the helix API before, the kraken API is depracated and will be removed at some future date.

**Review by mikeshardmind — APPROVED** [review]
Re-approving as this was just a merge for conflict resolution.

**Pull request merged**